### PR TITLE
Move serialize/deserialize to store config

### PR DESCRIPTION
**Pull request by jdlehman**
This moves serialize/deserialize out of the lifecycle and to the store config.

Issue #198

### CODE REVIEW

**Comment by jdlehman** [comment]
Still need to get the tests passing.  Looks like there might also be a problem with the store config object from prior commits.  There is a a Store that does not have the config set correctly.  Looks like it is coming from `store-model-test.js` and the store is set as:

```js
{ boundListeners: [ 'hello' ],
  StoreModel:
   { state: { x: 0 },
     bindListeners: { hello: [Object] },
     hello: [Function: hello],
     config: { getState: [Function: getState], setState: [Function: assign] } },
  dispatchToken: 'ID_41',
  displayName: '' }
```

which shows it is missing its config obj (it is on the StoreModel, but does not get set to the store correctly like the other alt Stores)

File: test/index.js [hunk 1]
```diff
@@ -208,6 +208,15 @@ class SecondStore {
 const secondStore = alt.createStore(SecondStore, 'AltSecondStore')
 
 class LifeCycleStore {
+  static config = {
+    serialize: (state) => {
+      state.serialized = true
```
Comment (goatslacker) [inline test/index.js]: return from here?

**Comment by goatslacker** [comment]
Fixed the tests, we were just looking in the wrong place for the config. There's one more test that needs fixing and it's the app-level serialization/deserialization configuration.

What's currently happening:

* When a store is created it uses the app-level configuration as a base: https://github.com/goatslacker/alt/blob/master/src/alt/index.js#L65
* This is so we can override getState, setState for every store without having to pass it individually to each store: https://github.com/goatslacker/alt/blob/master/src/alt/utils/createStoreConfig.js#L3

The problem is that we now have serialize/deserialize hooks that determine how a store should be serialized/deserialized. But we also have app-level serialize/deserialize if you want to use YAML instead of JSON for example.

Solutions?

File: test/index.js (hunk 1 above)
Comment (jdlehman) [inline test/index.js]: whoops haha

**Comment by jdlehman** [comment]
Hmm I did notice that and thought it might end up being a problem.  I guess we could just rename one pair.  I think serialize/deserialize makes sense at the app-level, as that is what JSON.stringify/JSON.parse are doing.  At the model level, it is pretty much selecting or setting up the data it wants to serialize/deserialize.  We could call them `preSerialize` or `serializePrepare` or `dataForSerialization`?

Just ideas off the top of my head.

**Comment by jdlehman** [comment]
Yep, that fixes them. Once we settle on a name, I can get this up and ready to merge.

**Comment by goatslacker** [comment]
how about onSerialize and onDeserialize?

**Comment by jdlehman** [comment]
Oh, I did `serializeData` and `deserializeData` just to make sure it worked. I will s/serializeData/onSerialize and we can merge.

File: docs/serialization.md [hunk 2]
```diff
@@ -7,17 +7,17 @@ permalink: /docs/serialization/
...
 
-## Serialize
+## onSerialize
 
-`serialize` provides a hook to transform the store data (via an optional return value) to be saved to an alt snapshot. If a return value is provided than it becomes the value of the store in the snapshot. For example, if the store name was `MyStore` and `serialize` returned `{firstName: 'Cereal', lastName: 'Eyes'}`, the snapshot would contain the data `{...'MyStore': {'firstName': 'Cereal', 'lastName': 'Eyes'}...}`. If there is no return value, the default, [`MyStore#getState()`](stores.md#storegetstate) is used for the snapshot data.
+`onSerialize` provides a hook to transform the store data (via an optional return value) to be saved to an alt snapshot. If a return value is provided than it becomes the value of the store in the snapshot. For example, if the store name was `MyStore` and `onSerialize` returned `{firstName: 'Cereal', lastName: 'Eyes'}`, the snapshot would contain the data `{...'MyStore': {'firstName': 'Cereal', 'lastName': 'Eyes'}...}`. If there is no return value, the default, [`MyStore#getState()`](stores.md#storegetstate) is used for the snapshot data.
 
-## Deserialize
+## DeonSerialize
```
Comment (goatslacker) [inline docs/serialization.md]: :fire:

Comment (jdlehman) [inline docs/serialization.md]: NOOOO :sob:

File: src/utils/ImmutableUtil.js [hunk 3]
```diff
@@ -48,6 +30,14 @@ function immutable(store) {
 
     getState(currentState) {
       return currentState
+    },
+
+    onSerialize: (state) => {
```
Comment (goatslacker) [inline src/utils/ImmutableUtil.js]: either lets make these all arrow functions or lets keep them as method shorthand.

I vote for arrow:

```js
getState: state => state,
onSerialize: state => state.toJS(),
...
```

**Comment by goatslacker** [comment]
:+1:

File: docs/serialization.md (hunk 2 above)
Comment (jdlehman) [inline docs/serialization.md]: Realized after that file that I needed to substitute deserialize first, but I missed fixing the mistakes with the capital D. #vimproblems

**Pull request merged**